[WIP] Rewrote the URL dispatcher API.

knbk opened the pull request.

collinanderson commented on django/core/urlresolvers.py:
```diff
@@ -1,640 +1,10 @@
+    "django.core.urlresolvers is deprecated and will be removed"
+    "in Django 2.0. Use django.urls instead.",
+    RemovedInDjango20Warning, stacklevel=2
+)
```
I personally wouldn't mind if we didn't deprecate this, and left these shims in a bit longer. (Makes it just a hair easier for people upgrading)

carljm commented on django/urls/constraints.py:
```diff
@@ -0,0 +1,135 @@
+
+    def describe(self):
+        pattern = self.regex.pattern
+        return str(pattern[1:]) if pattern.startswith('^') else str(pattern)
```
Why should the description hide the fact that a regex is start-anchored?

knbk replied:
Historically the different regexes for includes and views are concatenated to form a single regex in all error messages and the like, and only the first start anchor is kept (see `BaseResolver.describe()`). At the moment it's a bit of an ad hoc implementation, and it doesn't account for the increased complexity of Constraints yet.

carljm commented:
Hi @knbk -- thanks for all your work on this! What's the current status, in your mind? Ready for review? Still needs work? If the answer is "not really ready for review yet," I'm happy just to wait. If you do feel it's ready for review and testing, it would be helpful to get a prose overview of the PR -- what it changes, why, and what the user-facing impact and benefit is. That could be as an email to django-developers, or as some docs added to this PR, or both.

knbk commented:
Hi @carljm -- I have spent most of the sprints thinking about the internal structure of the resolvers in relation to reversing urls. I aim to implement this over the next few weeks. The `describe()` API still needs a bit of work as well to account for possible new Constraints which don't follow the current pattern. After that I'll put the API up for review on the mailing list.

auvipy commented:
@knbk ready for review?

knbk commented:
@auvipy Almost, there are a few more API changes I'd like to make, but nothing particularly complex. I think I'll have it ready for review somewhere this week.

charettes commented on django/core/checks/urls.py:
```diff
@@ -14,17 +14,19 @@ def check_resolver(resolver):
-            warnings.extend(check_pattern_name(pattern))
+    # We only check when we're using a built-in resolver
+    if isinstance(resolver, Resolver):
+        for resolver in resolver.resolvers:
```
I would avoid cloaking the `resolver` variable, I like your `sub_resolver` naming above.

knbk commented:
buildbot, test this please

charettes commented on tests/urlpatterns_reverse/tests.py:
```diff
@@ -912,7 +907,6 @@ def test_no_handler_exception(self):
 
 @override_settings(ROOT_URLCONF='urlpatterns_reverse.namespace_urls')
 class ResolverMatchTests(SimpleTestCase):
-
```
Unnecessary change.

timgraham commented:
To make the review more manageable, how about doing the urlresolvers move as a separate pull request that precedes the new feature additions?

gst commented on django/conf/urls/__init__.py:
```diff
@@ -59,27 +59,108 @@ def include(arg, namespace=None, app_name=None):
+        """
+        func = self.view
+        if isinstance(func, functools.partial):
+            func = func.func
```
regarding what's following : 
what if `func.func` is another partial (etc) ; I think it could be but I could be wrong both in that and also in my question (you could want to use the partial class name, though I don't think so) ?

knbk replied:
This is copied from [`RegexURLPattern.lookup_str()`](https://github.com/django/django/blob/dbb0df2a0ec5bee80bee336fc81408efb30b7e47/django/core/urlresolvers.py#L217). If you want you can change it upstream, then I'll change it here as well.

gst replied:
didn't knew about RegexURLPattern having the same method..
reading the docstring I'm thinking both should do this:
```python
# ..
if isinstance(func, functools.partial):
    while isinstance(func, functools.partial):
        func = func.func
# ...
```
shouldn't it ?
again that's my wild guess here, I don't have read more about the context of how this method is used (nor the other one), etc..

gst replied:
hmmm.. now I think I'm wrong, you want to only have the first partial func.. `func.func` shouldn't anyway be another partial or else that would be problematic I think, as views have to return something else.. Sorry for disturbance ;)

charettes replied:
@gst, the initial `if` statement is not required, the `while` conditional should be enough.

This change seems to be required to flatten nested `functool.partial` (`functool.partial(functool.partial(view, foo=1), bar=2)`) on Python < 3.5 which [automatically flatten them](https://docs.python.org/dev/whatsnew/3.5.html).

But like Marteen said this should be tracked by another PR/ticket.

gst replied:
happy new year !
yeah effectively finally this is required for python < 3.5 as you explain (I just tested it in an interactive session).
I'm happy to open a PR with this specific change so.. should not be too long to come..

gst replied:
hi @charettes ,
I've a branch, based on master, with a test case (test_data in urlpatterns_reverse.test) in its own commit and the fix in a second commit, could/should I open a PR with "only" that or is there a need to have also a ticket on its own for this ?

timgraham replied:
Please open a ticket to track the bug (all non-trivial changes should have a ticket).

akaariai commented:
I've ran djangobench on this branch, the results are that url_resolve is around 20% slower than on master, and url_reverse is around 30% slower. The reverse case is more interesting as url reversing tends to be done multiple times per requests in many common cases (JSON response containing a list of elements, each of the elements containing an URL for example).

knbk commented:
@akaariai I've made some changes to `reverse()`. Quoting string was by far the most expensive operation, so I moved it around a bit and removed a few redundant calls to `quote()`. Reversing now seems to be about the same, and even a bit faster on python 2.

apollo13 commented on tests/urls_tests/test_resolvers.py:
```diff
@@ -0,0 +1,239 @@
+from .views import class_based_view, empty_view
+
+
+@skip("Refactored interals.")
```
Skip?

knbk replied:
These tests are producing errors because they depend on internals of previous iterations. The refactor changed those internals. I'll have to take some time to fix them.

knbk commented:
I've separated the move to `django.urls` into a new PR: https://github.com/django/django/pull/5898

aaugustin commented:
Now that #5898 has been merged, rebasing on top of master would make reviewing easier.

knbk commented:
buildbot, test this please

aaugustin commented on django/conf/urls/__init__.py:
```diff
@@ -17,6 +16,8 @@
 
 
 def include(arg, namespace=None, app_name=None):
```
Is there a reason to keep the URL-related APIs split between `django.conf.urls` and `django.urls`?

I believe everything should be moved to `django.urls`.

knbk replied:
@aaugustin My initial idea was that third-party URL dispatchers may choose to use `django.conf.urls`, so I kept it as a separate module. I don't have a particular strong opinion on this, so if you feel that it should still be moved to `django.urls`, I'm happy to make that change.

aaugustin replied:
If we believe that most users of Django won't be able to reverse-engineer or remember the reason for having two distinct modules — and I do — then we should put everything in the same module.

auvipy replied:
@knbk I do think keeping everything related to url under a singe module does make the most sense as Aymeric say's.

apollo13 commented:
@knbk Mhm, having a hard time reproducing the segfault, running in a loop in screen now and hoping that I get something.

apollo13 commented:
buildbot, test this please.

apollo13 commented:
Got a core at last:
```
(gdb) info threads
  Id   Target Id         Frame 
  2    Thread 0x7ff3584fc740 (LWP 23161) vgetargskeywords.35607 (args=(), keywords=<optimized out>, format=0x5baea4 "iO:acquire", kwlist=<optimized out>, p_va=p_va@entry=0x7ffeb0d821e0, flags=flags@entry=0)
    at ../Python/getargs.c:1628
* 1    Thread 0x7ff33e53f700 (LWP 28310) __GI___libc_free (mem=0x38007600) at malloc.c:2929
(gdb) bt
#0  __GI___libc_free (mem=0x38007600) at malloc.c:2929
#1  0x00007ff354bd2adc in ffi_call_unix64 () from /usr/lib/x86_64-linux-gnu/libffi.so.6
#2  0x00007ff354bd240c in ffi_call () from /usr/lib/x86_64-linux-gnu/libffi.so.6
#3  0x00007ff354ec55db in _call_function_pointer (argcount=1, resmem=0x7ff33e53e550, restype=<optimized out>, atypes=<optimized out>, avalues=0x7ff33e53e530, pProc=0x7ff35184b1c0 <finishGEOS_r>, flags=4353)
    at /build/python3.5-AMbzgl/python3.5-3.5.1/Modules/_ctypes/callproc.c:811
#4  _ctypes_callproc (pProc=pProc@entry=0x7ff35184b1c0 <finishGEOS_r>, argtuple=argtuple@entry=(939554304,), flags=4353, argtypes=argtypes@entry=0x0, 
    restype=restype@entry=<_ctypes.PyCSimpleType at remote 0x12a18e8>, checker=0x0) at /build/python3.5-AMbzgl/python3.5-3.5.1/Modules/_ctypes/callproc.c:1149
#5  0x00007ff354ec59b8 in PyCFuncPtr_call (self=<optimized out>, inargs=<optimized out>, kwds=0x0) at /build/python3.5-AMbzgl/python3.5-3.5.1/Modules/_ctypes/_ctypes.c:3869
#6  0x000000000050578a in PyObject_Call (func=<_FuncPtr(__name__='finishGEOS_r') at remote 0x7ff33e8bbc00>, arg=<optimized out>, kw=<optimized out>) at ../Objects/abstract.c:2165
#7  0x0000000000568210 in do_call (nk=<optimized out>, na=<optimized out>, pp_stack=0x7ff33e53e830, func=<_FuncPtr(__name__='finishGEOS_r') at remote 0x7ff33e8bbc00>) at ../Python/ceval.c:4887
#8  call_function (oparg=<optimized out>, pp_stack=0x7ff33e53e830) at ../Python/ceval.c:4683
#9  PyEval_EvalFrameEx (
    f=f@entry=Frame 0x7ff340042438, for file /home/jenkins/workspace/pull-requests-trusty/database/spatialite/label/trusty-pr/python/python3.5/django/contrib/gis/geos/libgeos.py, line 147, in __del__ (self=<GEOSContextHandle(ptr=939554304) at remote 0x7ff33f8e9470>), throwflag=throwflag@entry=0) at ../Python/ceval.c:3185
#10 0x0000000000565f4e in _PyEval_EvalCodeWithName (_co=<code at remote 0x7ff353a64150>, globals=<optimized out>, locals=<optimized out>, args=<optimized out>, argcount=1, kws=0x0, kwcount=kwcount@entry=0, 
    defs=defs@entry=0x0, defcount=defcount@entry=0, kwdefs=kwdefs@entry=0x0, closure=closure@entry=0x0, name=name@entry=0x0, qualname=qualname@entry=0x0) at ../Python/ceval.c:3966
#11 0x000000000056eee8 in PyEval_EvalCodeEx (_co=<optimized out>, globals=<optimized out>, locals=<optimized out>, args=<optimized out>, argcount=<optimized out>, kws=<optimized out>, kwcount=0, defs=0x0, 
    defcount=0, kwdefs=0x0, closure=0x0) at ../Python/ceval.c:3987
#12 0x000000000056ef8c in function_call.86802 (func=<function at remote 0x7ff353a3bae8>, arg=(<GEOSContextHandle(ptr=939554304) at remote 0x7ff33f8e9470>,), kw=0x0) at ../Objects/funcobject.c:632
#13 0x000000000050578a in PyObject_Call (func=<function at remote 0x7ff353a3bae8>, arg=<optimized out>, kw=<optimized out>) at ../Objects/abstract.c:2165
#14 0x00000000004b450d in method_call.83066 (func=<function at remote 0x7ff353a3bae8>, arg=(<GEOSContextHandle(ptr=939554304) at remote 0x7ff33f8e9470>,), kw=0x0) at ../Objects/classobject.c:330
#15 0x000000000050578a in PyObject_Call (func=<method at remote 0x7ff3430575c8>, arg=<optimized out>, kw=<optimized out>) at ../Objects/abstract.c:2165
#16 0x0000000000505847 in PyEval_CallObjectWithKeywords (func=func@entry=<method at remote 0x7ff3430575c8>, arg=(), arg@entry=0x0, kw=kw@entry=0x0) at ../Python/ceval.c:4530
#17 0x00000000004961bd in slot_tp_finalize.23456 (self=<optimized out>) at ../Objects/typeobject.c:6330
#18 0x0000000000583f64 in PyObject_CallFinalizer (self=self@entry=<GEOSContextHandle(ptr=939554304) at remote 0x7ff33f8e9470>) at ../Objects/object.c:294
#19 0x0000000000539046 in PyObject_CallFinalizerFromDealloc (self=self@entry=<GEOSContextHandle(ptr=939554304) at remote 0x7ff33f8e9470>) at ../Objects/object.c:311
#20 0x00000000004b2910 in subtype_dealloc.22464 (self=<GEOSContextHandle(ptr=939554304) at remote 0x7ff33f8e9470>) at ../Objects/typeobject.c:1164
#21 0x000000000058630f in free_keys_object.16200 (keys=0x7ff338061d40) at ../Objects/dictobject.c:354
#22 0x000000000057be80 in dict_dealloc.16256 (mp=0x7ff34024c488) at ../Objects/dictobject.c:1604
#23 0x0000000000511c22 in PyDict_DelItem (
    op={<weakref at remote 0x7ff353a63098>: {'ptr': <LP_GEOSContextHandle_t at remote 0x7ff33f14b598>, 'handle': <GEOSContextHandle(ptr=<...>) at remote 0x7ff33f02c6a0>}}, 
    key=key@entry=<weakref at remote 0x7ff33fdd4ae8>) at ../Objects/dictobject.c:1290
#24 0x00000000004572b4 in _localdummy_destroyed.48186 (localweakref=<optimized out>, dummyweakref=<weakref at remote 0x7ff33fdd4ae8>) at ../Modules/_threadmodule.c:967
#25 0x00000000004b2e1c in PyCFunction_Call (func=<built-in method _localdummy_destroyed of weakref object at remote 0x7ff353a63048>, args=(<weakref at remote 0x7ff33fdd4ae8>,), kwds=<optimized out>)
    at ../Objects/methodobject.c:134
#26 0x000000000050578a in PyObject_Call (func=<built-in method _localdummy_destroyed of weakref object at remote 0x7ff353a63048>, arg=<optimized out>, kw=<optimized out>) at ../Objects/abstract.c:2165
#27 0x00000000004f7d3c in PyObject_CallFunctionObjArgs (callable=callable@entry=<built-in method _localdummy_destroyed of weakref object at remote 0x7ff353a63048>) at ../Objects/abstract.c:2445
#28 0x0000000000496143 in handle_callback.28519 (ref=ref@entry=0x7ff33fdd4ae8, callback=callback@entry=<built-in method _localdummy_destroyed of weakref object at remote 0x7ff353a63048>)
    at ../Objects/weakrefobject.c:868
#29 0x00000000004b0faa in PyObject_ClearWeakRefs (object=object@entry=<_thread._localdummy at remote 0x7ff3401ece10>) at ../Objects/weakrefobject.c:913
#30 0x00000000004b14c0 in localdummy_dealloc.48152 (self=0x7ff3401ece10) at ../Modules/_threadmodule.c:608
#31 0x000000000058630f in free_keys_object.16200 (keys=0x7ff338061e30) at ../Objects/dictobject.c:354
#32 0x000000000057be80 in dict_dealloc.16256 (mp=0x7ff33f148148) at ../Objects/dictobject.c:1604
#33 0x0000000000578537 in PyThreadState_Clear (tstate=tstate@entry=0x378e8c0) at ../Python/pystate.c:378
#34 0x000000000058c298 in t_bootstrap.48092 (boot_raw=0x2c15ef0) at ../Modules/_threadmodule.c:1026
#35 0x00007ff3580d6182 in start_thread (arg=0x7ff33e53f700) at pthread_create.c:312
#36 0x00007ff357e0347d in clone () at ../sysdeps/unix/sysv/linux/x86_64/clone.S:111
(gdb) f 9
#9  PyEval_EvalFrameEx (
    f=f@entry=Frame 0x7ff340042438, for file /home/jenkins/workspace/pull-requests-trusty/database/spatialite/label/trusty-pr/python/python3.5/django/contrib/gis/geos/libgeos.py, line 147, in __del__ (self=<GEOSContextHandle(ptr=939554304) at remote 0x7ff33f8e9470>), throwflag=throwflag@entry=0) at ../Python/ceval.c:3185
3185	../Python/ceval.c: No such file or directory.
(gdb) py-bt
Traceback (most recent call first):
  File "/home/jenkins/workspace/pull-requests-trusty/database/spatialite/label/trusty-pr/python/python3.5/django/contrib/gis/geos/libgeos.py", line 147, in __del__
    lgeos.finishGEOS_r(self.ptr)
  <built-in method _localdummy_destroyed of weakref object at remote 0x7ff353a63048>
(gdb) f 5
#5  0x00007ff354ec59b8 in PyCFuncPtr_call (self=<optimized out>, inargs=<optimized out>, kwds=0x0) at /build/python3.5-AMbzgl/python3.5-3.5.1/Modules/_ctypes/_ctypes.c:3869
3869	/build/python3.5-AMbzgl/python3.5-3.5.1/Modules/_ctypes/_ctypes.c: No such file or directory.
(gdb) py-print self
local 'self' = <GEOSContextHandle(ptr=939554304) at remote 0x7ff33f8e9470>
```

apollo13 commented:
buildbot, test this please.

apollo13 commented:
buildbot, test this please.

timgraham commented:
Marten, could you try reverting 312fc1af7b8a028611b45ee272652d59ea7bd85a in your branch and see if the segfault goes away?

knbk commented:
buildbot, test this please

apollo13 commented:
buildbot, test this please.

apollo13 commented:
buildbot, test this please.

knbk commented:
buildbot, test this please

jaddison commented:
Hi Marten - have you got any status updates on this PR? Would love to know if you're still able to make it for feature freeze?

Also, I submitted a docs-related PR against your branch a couple of weeks ago, wondering if you had a chance to review?

knbk commented:
Hi James, I'm sorry I didn't follow through on our conversation. I'm at DjangoCon in Budapest right now, I will definitely work on this at the sprints. I'm still aiming for 1.10. It will require some legwork, but I think it's manageable. 

I've taken a quick look at your docs PR, and it seems good at first sight. I'll take a closer look at the sprints (Saturday and Sunday), and merge it.

jaddison commented:
Marten, I'm not sure where to create issues against your 'dispatcher_api' branch, so I'll add it here: as a reminder to our discussion on IRC a month or so ago (see http://django-irc-logs.com/2016/mar/09/#2221937), redirecting to different sub-domains is prevented. So logging in on `https://secure.example.com/user/login/?next=http://www.example.com/my-cool-page/` doesn't work.

This is even with settings like the following:

```
ALLOWED_HOSTS = ['.example.com']
```

It's `is_safe_url()` that doesn't allow redirection between different domains.  See https://github.com/knbk/django/blob/dispatcher_api/django/contrib/auth/views.py#L122, which only validates the URL to redirect to against the *current* host/domain name.

jaddison commented:
I thought of something else that could be problematic with multiple subdomain configurations - `request. build_absolute_uri(location)`, if provided with a `location` URL existing on a separate subdomain in the project's URL patterns.

Similiar to the redirection issue I mentioned in my previous comment, `request.build_absolute_uri()` also uses `request.get_host()`, which itself uses only the current request's domain.

auvipy commented:
@knbk hi, is this not going to be merged on 1.10? how much work is remain?

timgraham commented:
Closing for now since Marten isn't activity working on this. We can reopen later or start a new PR.

timgraham commented:
Is there a ticket for this? It would be nice to have one so the work doesn't get lost.
